### PR TITLE
A basic implementation of mock TiKV API surface

### PR DESCRIPTION
**Pull request by sunxiaoguang**
Trying to split the initial working version to several smaller commits. This is the first one with definition and implementation of basic mock TiKV API

### CODE REVIEW

File: server/api/store.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+// limitations under the License.
+
+package api
+
+// StoreState contains state about a store.
+type StoreState string
+
+const (
+	// StoreStateUp store is up and running
+	StoreStateUp = "Up"
```
Comment (rleungx) [inline server/api/store.go]: How about `StoreStateUp StoreState = "Up"`?

File: server/cluster.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	clusterID uint64
+
+	member  *memberInstance // we only support one leader pd at this time
+	regions []*regionInstance
+	stores  []*storeInstance
+
+	idAlloc uint64
+}
+
+func (s *Server) doGetCluster(id uint64) *clusterInstance {
```
Comment (rleungx) [inline server/cluster.go]: Can we just use `GetCluster`?

File: server/cluster.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+			if len(region.StartKey) != 0 {
+				return errors.New(fmt.Sprintf("Start key of first region '%d' isn't empty", region.Id))
+			}
+			if region.EndKey == nil {
+				region.EndKey = emptyKey
+			}
+		} else {
+			if !bytes.Equal(region.StartKey, lastEndKey) {
+				return errors.New(fmt.Sprintf("Start key '%s' of region '%d' isn't the same as end key '%s' of last region '%d'",
+					base64.StdEncoding.EncodeToString(region.StartKey), region.Id,
```
Comment (rleungx) [inline server/cluster.go]: Do we need to use hex instead of base64 here to keep consistent with PD or TiKV?

File: server/region.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	return &regionInstance{
+		Region: metapb.Region{
+			Id:       id,
+			StartKey: startKey,
+			EndKey:   endKey,
+		},
+	}
+}
+
+func (r *regionInstance) start(idAlloc idAllocator, store *storeInstance) error {
```
Comment (rleungx) [inline server/region.go]: Why does `regionInstance` have a `start` function?

File: server/server.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+// Server is the mock tikv server.
+type Server struct {
+	// Configs and initial fields.
+	cfg *Config
+
+	address string
+	ctx     context.Context
+	cancel  context.CancelFunc
+
+	idAlloc  uint64
```
Comment (rleungx) [inline server/server.go:50]: Is it never used?

File: server/cluster.go (hunk 2 above)
Comment (sunxiaoguang) [reply server/cluster.go]: I was trying to put all glue code between server and api in handler.go and conversion.go. So there are some method in handler being very simple and basically just passing arguments to server implementation. I used getCluster initially but revive complains that these two methods are confusing. After scratching my head I came up with these not good names. So consider this isn't going to be public accessible and should start with lowercase what could be a better candidate?

File: server/cluster.go (hunk 3 above)
Comment (sunxiaoguang) [reply server/cluster.go]: Sure, let me change it.

File: server/server.go (hunk 5 above)
Comment (sunxiaoguang) [reply server/server.go:50]: Yes, it's in a very early stage and many things are just there for future.

File: server/region.go (hunk 4 above)
Comment (sunxiaoguang) [reply server/region.go]: Right this name is confusing. Let's figure out a better name.

**Comment by brson** [comment]
What's next for this pr @rleungx?

**Comment by siddontang** [comment]
@brson 

@disksing and @sunxiaoguang have discussed this before, we decide to do:

1. move all associated codes from tidb to client-go
2. extract the mock TiKV codes from client-go
3. add gRPC support to mock TiKV so other no Go clients can use it

File: server/api/store.go (hunk 1 above)
Comment (sunxiaoguang) [reply server/api/store.go]: Fixed

File: server/cluster.go (hunk 3 above)
Comment (sunxiaoguang) [reply server/cluster.go]: Fixed

File: server/region.go (hunk 4 above)
Comment (sunxiaoguang) [reply server/region.go]: Changed to `init`

**Comment by sunxiaoguang** [comment]
@rleungx @brson @siddontang @disksing 
I pushed some more changes to this PR to include a functional mock TiKV cluster with gRPC support. client-rust can already run tests against this mock TiKV implementation. Can we setup an online meeting or onsite discussion after Chinese New Year holiday to discuss something more about this project?

File: cmd/mock-tikv/main.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/grpc-ecosystem/go-grpc-prometheus"
+	"github.com/pingcap/pd/pkg/logutil"
+	"github.com/pingcap/pd/pkg/metricutil"
```
Comment (disksing) [inline cmd/mock-tikv/main.go]: Better not to depend on pd project. I prefer copy the utils.

File: server/cluster.go [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+	stores     []*storeInstance
+	storeByID  map[uint64]*storeInstance
+
+	idAlloc uint64
+
+	store *MVCCLevelDB
+}
+
+func validateRegionAndStore(regions []*regionInstance, stores []*storeInstance) error {
+	var lastEndKey []byte = nil
```
Comment (disksing) [inline server/cluster.go]: No need `= nil`, the default value is alraedy nil.

File: server/cluster.go [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+	regionByID map[uint64]*regionInstance
+	stores     []*storeInstance
+	storeByID  map[uint64]*storeInstance
+
+	idAlloc uint64
+
+	store *MVCCLevelDB
+}
+
+func validateRegionAndStore(regions []*regionInstance, stores []*storeInstance) error {
```
Comment (disksing) [inline server/cluster.go]: How `stores` is validated?

File: server/cluster.go [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+	if len(lastEndKey) != 0 {
+		return errors.New(fmt.Sprintf("Last end key '%s' of region '%d' isn't empty",
+			hex.EncodeToString(lastEndKey), lastRegion.Id,
+		))
+	}
+	return nil
+}
+
+func (c *clusterInstance) startStores(address string) error {
+	for _, store := range c.stores {
```
Comment (disksing) [inline server/cluster.go:93]: Read lock?

File: cmd/mock-tikv/main.go (hunk 6 above)
Comment (sunxiaoguang) [reply cmd/mock-tikv/main.go]: Fixed

File: server/cluster.go (hunk 9 above)
Comment (sunxiaoguang) [reply server/cluster.go:93]: The connection information isn't returned therefore no client can connect to it at this time.

File: server/cluster.go (hunk 8 above)
Comment (sunxiaoguang) [reply server/cluster.go]: Hm, at this time it only verifies key range of regions.  Let me change it to validateRegion and remove stores argument.

File: server/cluster.go (hunk 7 above)
Comment (sunxiaoguang) [reply server/cluster.go]: Fixed

**Comment by sunxiaoguang** [comment]
@disksing PTAL

**Review by disksing — APPROVED** [review]

File: server/api/server.go [hunk 10]
```diff
@@ -0,0 +1,48 @@
+// Copyright 2016 PingCAP, Inc.
```
Comment (rleungx) [inline server/api/server.go]: We need to change all licenses to 2019.

File: cmd/mock-tikv/main.go [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+	switch errors.Cause(err) {
+	case nil:
+	case flag.ErrHelp:
+		os.Exit(0)
+	default:
+		log.Fatalf("parse cmd flags error: %s\n", fmt.Sprintf("%+v", err))
+	}
+
+	err = server.InitLogger(&cfg.Log)
```
Comment (rleungx) [inline cmd/mock-tikv/main.go:43]: Do we need to use `pingcap/log` instead of `logrus`? /cc @disksing

Comment (disksing) [reply cmd/mock-tikv/main.go:43]: I think pingcap/log is fine. @rleungx

Comment (siddontang) [reply cmd/mock-tikv/main.go:43]: pingcap/log +1

It is faster than logrus.

Comment (sunxiaoguang) [reply cmd/mock-tikv/main.go:43]: Sure. Store fail points code is almost done. Let me cleanup code and fix these all together.

Comment (sunxiaoguang) [reply cmd/mock-tikv/main.go:43]: Changed to pingcap/log

File: server/api/server.go (hunk 10 above)
Comment (sunxiaoguang) [reply server/api/server.go]: Done

**Pull request merged**